src/sentry/api/endpoints/user_authenticator_enroll.py

```diff
@@ -0,0 +1,195 @@
+
+        # User is not enrolled in auth interface:
+        # - display configuration form
+        response = serialize(interface)
+        response['form'] = get_serializer_field_metadata(
+            serializer_map[interface_id]
+        )
+
+        # U2fInterface has no 'secret' attribute
+        try:
+            response['secret'] = interface.secret
+        except AttributeError:
+            pass
+
+        if interface_id == 'totp':
+            response['qrcode'] = interface.get_provision_qrcode(user.email)
+
+        if interface_id == 'u2f':
+            response['challenge'] = interface.start_enrollment()
+
+        return Response(response)
+
+    @sudo_required
+    def post(self, request, user, interface_id):
+        """
+        Enroll in authenticator interface
+        `````````````````````````````````
+
+        :pparam string user_id: user id or "me" for current user
+        :pparam string interface_id: interface id
+
+        :auth: required
+        """
+
+        # start activation
+        serializer_cls = serializer_map.get(interface_id, None)
+
+        if serializer_cls is None:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        serializer = serializer_cls(data=request.DATA)
+
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        interface = Authenticator.objects.get_interface(user, interface_id)
+
+        # Not all interfaces allow multi enrollment
+        #
+        # This is probably un-needed because we catch
+        # `Authenticator.AlreadyEnrolled` when attempting to enroll
+        if interface.is_enrolled and not interface.allow_multi_enrollment:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        try:
+            interface.secret = request.DATA['secret']
+        except KeyError:
+            pass
+
+        # Need to update interface with phone number before validating OTP
+        if 'phone' in request.DATA:
+            interface.phone_number = serializer.data['phone']
+
+            # Disregarding value of 'otp', if no OTP was provided,
+            # send text message to phone number with OTP
+            if 'otp' not in request.DATA:
+                if interface.send_text(for_enrollment=True, request=request._request):
+                    return Response(status=status.HTTP_204_NO_CONTENT)
+                else:
+                    # Error sending text message
+                    return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        # Attempt to validate OTP
+        if 'otp' in request.DATA and not interface.validate_otp(serializer.data['otp']):
+            return Response({'detail': 'Invalid OTP'}, status=status.HTTP_400_BAD_REQUEST)
+
+        # Try u2f enrollment
+        if interface_id == 'u2f':
+            # What happens when this fails?
+            interface.try_enroll(
+                request.DATA.get('challenge'),
```

Comment:
what will happen if these (challenge and response) are `None`? should you validate that they're not before calling this?